hls-test-utils/src/Test/Hls/FileSystem.hs

```diff
@@ -0,0 +1,231 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Test.Hls.FileSystem
+  ( FileSystem(..)
+  , VirtualFileTree(..)
+  , FileTree
+  , Content
+  -- * init
+  , materialise
+  , materialiseVFT
+  -- * Interaction
+  , readFileFS
+  , writeFileFS
+  -- * Test helpers
+  , mkVirtualFileTree
+  , toNfp
+  , toAbsFp
+  -- * Builders
+  , file
+  , copy
+  , directory
+  , text
+  , ref
+  -- * Cradle helpers
+  , directCradle
+  , simpleCabalCradle
+  -- * Full project setups
+  , directProject
+  , directProjectMulti
+  , simpleCabalProject
+  , simpleCabalProject'
+  ) where
+
+import           Data.Foldable               (traverse_)
+import qualified Data.Text                   as T
+import qualified Data.Text.IO                as T
+import           Development.IDE             (NormalizedFilePath)
+import           Language.LSP.Protocol.Types (toNormalizedFilePath)
+import           System.Directory
+import           System.FilePath             as FP
+
+-- ----------------------------------------------------------------------------
+-- Top Level definitions
+-- ----------------------------------------------------------------------------
+
+-- | Representation of a 'VirtualFileTree' that has been 'materialise'd to disk.
+--
+-- The 'fsTree' must be persisted in 'fsRoot'.
+data FileSystem =
+  FileSystem
+    { fsRoot         :: FilePath
+    , fsTree         :: [FileTree]
+    , fsTestDataRoot :: FilePath
+    } deriving (Eq, Ord, Show)
+
+-- | Virtual representation of a filesystem tree.
+--
+-- Operations of 'vftTree' are relative to 'vftTestDataRoot'.
+-- In particular, any 'copy' etc. operation looks for the sources in 'vftTestDataRoot'.
```

Comment:
This seems a little odd to me. At the moment, a `VirtualFileTree` is actually defined relative to some real files on disk. It _could_ be entirely in memory, with the contents of the files read in. It seems to me that you could have a pattern like this:
```
data OnDisk = OnDisk
data InMemory = InMemory T.Text

data FileTree a
  = File FilePath a
  | Directory FilePath [FileTree]
  deriving (Show, Eq, Ord)

data FileSystem = FileSystem { root :: FilePath, tree :: [FileTree OnDisk] }
data VirtualFileSystem = VirtualFileSystem { tree :: FileTree InMemory }

virtualizeFS :: FileSystem -> VirtualFileSystem
materializeFS :: FilePath -> VirtualFileSystem -> IO FileSystem
```
That is:
- User specifies the "real file system" in the test directory
- That gets "virtualized" into a VFS that has all the content in memory
- The VFS gets "materialized" back out into a real file system in a different root.

---

I wrote all this and then I looked further down and I see why this would be awkward. You want to mix the specification of files from disk with programmatically specified files when you specify the input file system. I'll leave this here as an idea, anyway. I think you could still maybe make it work if you had something like:
```
virtualizeFS (realFS [ "foo"
   directory "bar" [
      "baz.hs"
   ]
]) <> directCradle args

Reply:
Or maybe I just want a third type?
```
data FileSystemSpec = FileSystemSpec { root :: FilePath, tree :: [FileTree Content] }
virtualizeFS :: FileSystemSpec -> VirtualFileSystem
```

Reply:
Yeah, I agree, that would be all really cool. If we manage to avoid IO, then our tests will be incredibly fast!

Reply:
Oh, I wasn't actually suggesting we run on the VFS. Just that the current types seem to be a mix of the virtual and non-virtual that I find a bit confusing.

Reply:
In my opinion, this mix is an essential part, and still explicit enough.